### ISSUE TITLE
MRIBase.rawdata does not work correctly with shuffled profiles

### ISSUE THREAD

**Issue by atsanda** [MagneticResonanceImaging/MRIReco.jl#215]
In the current form `MRIBase.rawdata` operates under assumption that profiles come in a certain order which then allows to place readout data in certain places of the multidimensional `kdata` array. I find this assumption relatively strong and suggest lifting it. I am going to write a test suite to highlight the issue and to later fix the bug.

**Comment by atsanda** [comment]
Here is an example:
https://github.com/MagneticResonanceImaging/MRIReco.jl/blob/bug-issue-215-MRIBase.rawdata-shuffled-profiles/MRIBase/test/testRawAcqData.jl

**Comment by aTrotier** [comment]
Do you expects MRIBase.rawdata to return something ordered like a k-space ?

This is done at the next steps after the conversion from raw -> acq and then you can use kDataCart(acq::AcquisitionData)

**Comment by tknopp** [comment]
yes, `rawdata` should basically output the data in the way that it has been acquired by the MRI scanner. The resorting is done afterwards when converting to an AcquisitionData object.

**Comment by atsanda** [comment]
@aTrotier, yes, kspace should be ordered after that, I mean the encoded part of it. I've taken a look into kDataCart. I don't see how it could help solving the issue. It orders according to repetitions, slices, echos and channels but the data within profiles is just copied. Do I understand it right?

**Comment by aTrotier** [comment]
@atsanda MRIBase.rawdata is not suppose to do that.
The RawAcquisitionData structure is mostly a copy of the ISMRMRD standard which stores the data in the order they are acquired. The `rawdata` function just retrieve the rawdata for specific echo, slice and rep which right now are the 3 dimensions others than x/y/z which are used later by MRIReco

**Comment by aTrotier** [comment]
kDataCart is a function I wrote first to help me for conversion back and forth in order to call BartIO.jl
I think it is only used in espirit for simplicity.

kData is really just extracting the kdata for a certain k,j,i,rep 
```julia
"""
    kData(acqData::AcquisitionData, echo::Int64=1, coil::Int64=1, slice::Int64=1;rep::Int64=1)

returns the k-space contained in `acqData` for given `echo`, `coil`, `slice` and `rep`(etition).
"""
function kData(acqData::AcquisitionData, echo::Int64=1, coil::Int64=1, slice::Int64=1;rep::Int64=1)
  return acqData.kdata[echo,slice,rep][:,coil]
end
```

but the encoding operator F is used here to perform the FFT and subsample / put the corresponding k-space lines at the right place. The ordered k-space is not really created anywhere.

You're right the rawdata function is especially complicated but I am not sure we can get rid of the filter. We should add a test to see what happens if a line is present 2 times in the profiles. 

I guess in that case the data should be averages but in the current implementation we only use the first one

**Comment by aTrotier** [comment]
> Then it is just the output of the rawdata which is taken for the reconstruction. So, in this case, should I manually extract kspace with kDataCart and pass it to inverse fourier or is there a standard way of doing it?

```julia
      if !MRIBase.isUndersampledCartTrajectory(shape,tr)
        ftOp = FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      else
        idx = MRIBase.cartesianSubsamplingIdx(shape,tr)
        ftOp = SamplingOp(Complex{T}; pattern=idx, shape, S = S) ∘ FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      end
```

Humm, actually if the k-space is not detected as UnderSampledCartTrajectory I suppose it is a simple FFT operator

**Comment by atsanda** [comment]
@aTrotier 

> You're right the rawdata function is especially complicated but I am not sure we can get rid of the filter. We should add a test to see what happens if a line is present 2 times in the profiles.

I would expect it to fail due to dimensions mismatch with the Fourier operator. Data would have more points then than the operator would expect. I think the operator is allocated according to the trajectory which would contain the shape of the kspace. It is rather the question of reproducibility for this function. We can keep it simple and focused moving filters to the point where `AcqData` is created and where we expect the data to be transformed. Then it should work. I can prepare a pull request for that with tests if you agree that it is a good idea. I just need a point somewhere in the code where constructing a proper dense array for kspace with filters and actual indexes would be expected. If that's not the `rawdata`, would be great to find one. Otherwise the framework does not work for me.

> Humm, actually if the k-space is not detected as UnderSampledCartTrajectory I suppose it is a simple FFT operator
And that is my case, I just don't have some point around the y-z corners in my ksapce and I am fine with zeros there.

**Comment by aTrotier** [comment]
```julia
function isUndersampledCartTrajectory(shape::NTuple{3,Int}, tr::Trajectory)
  return shape[1]!=numSamplingPerProfile(tr) || shape[2]!=numProfiles(tr) || shape[3]!=numSlices(tr)
end
```

in your case the function returns true. can you remove the if part and always apply the samplingOp: 
```julia
      if !MRIBase.isUndersampledCartTrajectory(shape,tr)
        ftOp = FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      else
        idx = MRIBase.cartesianSubsamplingIdx(shape,tr)
        ftOp = SamplingOp(Complex{T}; pattern=idx, shape, S = S) ∘ FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      end
```
by
```julia
        idx = MRIBase.cartesianSubsamplingIdx(shape,tr)
        ftOp = SamplingOp(Complex{T}; pattern=idx, shape, S = S) ∘ FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)

```

@tknopp  Should we force the FFTOp to always be multiply by the SamplingOp ? Or do you want to keep the if/else and change how is define an undersampledCartTrajectory by something like :

```julia
idx = MRIBase.cartesianSubsamplingIdx(shape,tr) 
isUnderSample = length(idx) < prod(shape) ? true : false
```

**Comment by tknopp** [comment]
I would not always  perform the subsampling. It is an unnecessary operation for fully sampled data.

**Comment by aTrotier** [comment]
the conversion to AcquisitionData already filter the supplementary profiles (for cartesian acquisition at least) : 

```julia
using MRIReco
include("MRIBase/test/Utils.jl")

n_profiles = 10
n_samples_per_profile = 12
data = rand(ComplexF32, n_samples_per_profile, n_profiles)
f = Utils.get_default_cartesian_raw_acq_data_2d(n_profiles, n_samples_per_profile, data)

acq_1 = AcquisitionData(f)
length(acq_1.kdata[1,1,1]) == n_profiles * n_samples_per_profile

f2 = deepcopy(f)
push!(f2.profiles,f2.profiles[1])
acq_2 = AcquisitionData(f2)
length(acq_2.kdata[1,1,1]) == n_profiles * n_samples_per_profile
kDataCart(acq_2) == kDataCart(acq_1) # equal because only the first profile is used

f3 = deepcopy(f)
prof = deepcopy(f3.profiles[1])
prof.data .= 1
push!(f3.profiles,prof)
acq_3 = AcquisitionData(f3)
length(acq_3.kdata[1,1,1]) == n_profiles * n_samples_per_profile
kDataCart(acq_3) == kDataCart(acq_1) # equal because only the first profile is used 
kDataCart(acq_3) == kDataCart(acq_2) # equal because only the first profile is used 


f4 = deepcopy(f)
push!(f4.profiles,f4.profiles[1])
f4.profiles[1].data .= 1
acq_4 = AcquisitionData(f4)
length(acq_4.kdata[1,1,1]) == n_profiles * n_samples_per_profile
kDataCart(acq_4) == kDataCart(acq_1) # false because the first profile is used and is equal to 1
```

**Comment by atsanda** [comment]
Hm, sounds like a good reason to remove unnecessary filtration then. I will look into it and create a separate issue for refactoring.

Regarding the subsampling operator, it can help me with missing profiles but does not solve the issue with the ordering which I wanted to solve with this issue. As `kData` just takes the output of `rawdata` it is assumed that the order of profiles matches the memory layout of multidimensional kspace. It is not the case for me. `kDataCart` fixes that with proper re-indexing. I think we should integrate that into the creation of `AcqData` struct. it should not hurt the package as it makes it more general. That would be the second refactoring issue. @aTrotier , what would you say to that?

**Comment by aTrotier** [comment]
I still don't understand what is the issue with the indexing.

Yes, the kData are not reindexed (they are not supposed to). Even if we reordered them during the conversion to AcquisitionData, if the data are undersampled (with a CS or IPAT acceleration for example), the "order" kdata won't be useful. It will works only for a fully sampled kspace without acceleration, zero-filling, partial-fourier, cropped circle encoding.

**Comment by atsanda** [comment]
Well, first of all, this concrete issue, as discussed above, should be closed. `MRIBase.rawdata` is not supposed to do anything with profiles (although it does in fact). I wrote tests to document that. @aTrotier, could you, please, review the PR and merge? We could then close here with the `rawdata`.

However, the issue I have is still there. I have a 3D Cartesian measurement where kpace was not measured around the corners of ky-kz, a sort of ellipsoidal mask was applied along ky-kz. I try to reconstruct it with the framework and it fails. It happens because direct reconstruction accesses `rawdata` to apply inverse Fourier. The `rawdata` recognizes Cartesian trajectory, pre-allocates full zero-filled kspace for that and fills that in in the order profiles appear. In my case, that results in a complete corrupted kspace data. I expect missing parts to be filled with zeros and placed where they belong. I've boiled down this big issue to the problem of ordering. If at some point, `rawdata` or a function afterwards would not just take profiles the way they appear in the rawdata but took into account actual ky and kz values, my issue would be solved. It requires lifting the assumption that profiles come in the right order and there are enough profiles to fill a Cartesian grid.
I can also craft a toy ISMRMRD file to showcase my struggle. It well may be that the issue can be solved differently,

**Comment by aTrotier** [comment]
@atsanda I am closing this, Can you write an example to show this problem and open a new issue

**Comment by atsanda** [comment]
@aTrotier could you, please, review the associated PR with tests first?

**Comment by aTrotier** [comment]
> However, the issue I have is still there. I have a 3D Cartesian measurement where kpace was not measured around the corners of ky-kz, a sort of ellipsoidal mask was applied along ky-kz. I try to reconstruct it with the framework and it fails. It happens because direct reconstruction accesses rawdata to apply inverse Fourier. The rawdata recognizes Cartesian trajectory, pre-allocates a full zero-filled kspace for that and fills that in the order profiles appear. In my case, that results in a complete corrupted kspace data. I expect missing parts to be filled with zeros and placed where they belong. I've boiled down this big issue to the problem of ordering. If at some point rawdata or a function afterwards did not just take profiles the way they appear in the rawdata but took into account actual ky and kz values, my issue would be solved. It requires lifting the assumption that profiles come in the right order and there are enough profiles to fill a Cartesian grid.
I can also craft a toy ISMRMRD file to showcase my struggle. It well may be that the issue can be solved differently,


direct reconstruction is using the encodingOps_simple operators which use the trajectory in order perform the FFT. The reconstruction is suppose to works even with circular acquisition / compressed sensing and so on.

Maybe the issue is with the SamplingOp. Most of the time it is a problem with the subsampling indices during raw -> acq conversion

```julia
      @debug "FFTOp"
      if !MRIBase.isUndersampledCartTrajectory(shape,tr)
        ftOp = FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      else
        idx = MRIBase.cartesianSubsamplingIdx(shape,tr)
        ftOp = SamplingOp(Complex{T}; pattern=idx, shape, S = S) ∘ FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
      end
```

Can you take a look at `!MRIBase.isUndersampledCartTrajectory(shape,tr)` ?

---

However, I think their is an issue here :
```julia
      if !MRIBase.isUndersampledCartTrajectory(shape,tr)
        ftOp = FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
```

The FFTOp expects `rawdata(acq)` to be in the right order. I think we should remove the first part of the `if/else` and always use 
```julia
idx = MRIBase.cartesianSubsamplingIdx(shape,tr)
ftOp = SamplingOp(Complex{T}; pattern=idx, shape, S = S) ∘ FFTOp(Complex{T}; shape, unitary=false, S = S, fftParams(S)...)
```

The SamplingOp will put the profiles at the right position to perform the fft even for fully sampled with specific ordering acquisition. Are you ok if with that @tknopp @nHackel  ?
